src/main/java/org/kitteh/irc/client/library/event/client/ClientConnectionClosedEvent.java

```diff
@@ -24,40 +24,46 @@
 package org.kitteh.irc.client.library.event.client;
 
 import org.kitteh.irc.client.library.Client;
-import org.kitteh.irc.client.library.event.abstractbase.ClientEventBase;
 import org.kitteh.irc.client.library.util.ToStringer;
 
 import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Optional;
 
 /**
- * The {@link Client} has disconnected from the server.
+ * The {@link Client} has had a working connection cease.
  */
-public class ClientConnectionClosedEvent extends ClientEventBase {
-    private final boolean reconnecting;
+public class ClientConnectionClosedEvent extends ClientConnectionEndedEvent {
+    private final Optional<String> lastMessage;
 
     /**
      * Constructs the event.
      *
      * @param client client for which this is occurring
      * @param reconnecting true if the client plans to reconnect
+     * @param exception exception, if there was one, closing it
+     * @param lastMessage last message received
      */
-    public ClientConnectionClosedEvent(@Nonnull Client client, boolean reconnecting) {
-        super(client);
-        this.reconnecting = reconnecting;
+    public ClientConnectionClosedEvent(@Nonnull Client client, boolean reconnecting, @Nullable Exception exception, @Nullable String lastMessage) {
+        super(client, reconnecting, exception);
+        this.lastMessage = Optional.ofNullable(lastMessage);
```

Comment:
why isn't this wrapped in the getter? To [quote Brian Goetz](https://stackoverflow.com/a/26328555/2919551) about `Optional`:

> You should almost never use it as a field of something or a method parameter.

Reply:
Old, bad habit. Resolved soon.